Mender-server v4.0 upgrade docs (#2474)
mendersoftware/mender-docs: 07.Server-installation/02.Upgrading-from-previous-versions/091.Upgrading-from-3.7-to-4.0/docs.md

Reviewer: Should probably align this one with the version in `mender-helm` repository. I won't comment on things that I commented on there.

---
title: Upgrading to Mender Server 4.0
taxonomy:
category: docs
label: tutorial
---

This is a tutorial for upgrading the Mender Server from version 3.7 to
4.0 in production environments.

Besides the version specific guidelines below, please make sure you're following
the [suggested upgrade best practices](../docs.md) to reduce the risk of
unpredictable downtime and data loss.

## Before you start
Ensure you have the mender values file from the previous installation. 
You will need it as a base for the new installation:

<!--AUTOVERSION: "cp mender-%"/ignore -->
```bash

> cp mender-3.7.7.yml mender-values.yml

Reviewer: Where is `mender-3.7.7.yml` coming from?
```suggestion
cp mender-3.7*.yml mender-values.yml
```
I guess we could also assume the values could use the Helm version?
```suggestion
cp mender-5.*.yml mender-values.yml
```

Author: I don't think so: the `mender-3.7.7.yml` is coming from the [official 3.7 installation docs](https://docs.mender.io/3.7/server-installation/production-installation-with-kubernetes/mender-server#installing-the-mender-helm-chart). It's unlikely that somebody is using the mender helm chart version.

```

You can now start editing the `mender-values.yml` file with the following
changes.

### Current version

<!--AUTOVERSION: "Be sure that your current version is %"/ignore -->

> Be sure that your current version is 3.7 and that it works, and Mender Server starts properly.

Reviewer: Maybe we should include a mender-helm version requirement here as well (e.g. >= 5.9) so we don't get into some surprises?
I don't have an overview of which helm versions include 3.7, but there could be some potential unforeseen side-effects by relaxing the version requirements.

Reviewer: ```suggestion
Be sure that your current version is 3.7 (`mender-helm >= 5.9`) and that it works, and Mender Server starts properly.
```

Author: Yeah, you're right! Let's define the 5.9: it's not too old, and not the latest version, a bit more relaxed.

<!--AUTOVERSION: "## Migrating from %"/ignore -->
## Migrating from 3.7
The helm chart is versioned separately from the Mender releases.

<!--AUTOVERSION: "Until the helm chart version %"/ignore -->
<!--AUTOVERSION: "Mender % and %. Starting from Mender %, you have to use at least the"/ignore -->
<!--AUTOVERSION: "chart version %"/ignore -->
Until the helm chart version 5.x, the chart is compatible with both
Mender 3.6 and 3.7. Starting from Mender Server 4.0, you have to use at least the
chart version 6.0.x.
You should be aware of a few deprecations:

> <!--AUTOVERSION: "### Helm Chart v% deprecations:"/ignore -->
> ### Helm Chart v6.x.x deprecations

Reviewer: Heading with empty section. Consider removing it or replacing/merging with the parent heading.
```suggestion
```

#### Global section deprecations
The `global` section is internally dedicated to the global configuration,
for sub-charts also. So we are moving the Mender resources from the `global`
key to the `default` key. For this reason, the `global.image` key is 
moved to the `default.image` key: please make sure to comment out
your `global.image` key:
<!--AUTOVERSION: "  #   tag: mender-%"/ignore -->
```yaml
global:
  # image:
  #   tag: mender-3.7.7
```

#### New repositories location
Following a migration from multiple repositories to a single Monorepo, the
Container repositories has been moved to respect the new structure.
Additionally, the `mender-x.y` tag has been replaced with the `vx.y` tag.
* Enterprise registry:
  From `registry.mender.io/mendersoftware/deployments-enterprise:mender-3.7`
  to `registry.mender.io/mender-server-enterprise/deployments:v4.0`

* Open Source registry:
  From `docker.io/mendersoftware/deployments:mender-3.7`
  to `docker.io/mendersoftware/deployments:v4.0`

The new repository structure is already reflected in the default `values.yaml`

> file; please make sure to not override it in your `mender-values.yaml` file.

Reviewer: I believe anyone who has set up registry mirroring would appreciate a complete list of repository renaming.
Especially for enterprise, there are 3 different cases of renaming:
 1. Removing -enterprise suffix
 2. Changing `/mendersoftware` to `/mender-server-enterprise`
 3. Moving the remaining components from Docker Hub to `registry.mender.io`.

```markdown
[ui-tabs position="top-left" active="0" theme="default" ]
[ui-tab title="Open Source"]
| From                                                         | To                                                     |
| ------------------------------------------------------------ | ------------------------------------------------------ |
| docker.io/mendersoftware/create-artifact-worker:mender-X.Y.Z | docker.io/mendersoftware/create-artifact-worker:vX.Y.Z |
| docker.io/mendersoftware/deployments:mender-X.Y.Z            | docker.io/mendersoftware/deployments:vX.Y.Z            |
| docker.io/mendersoftware/deviceauth:mender-X.Y.Z             | docker.io/mendersoftware/deviceauth:vX.Y.Z             |
| docker.io/mendersoftware/deviceconfig:mender-X.Y.Z           | docker.io/mendersoftware/deviceconfig:vX.Y.Z           |
| docker.io/mendersoftware/deviceconnect:mender-X.Y.Z          | docker.io/mendersoftware/deviceconnect:vX.Y.Z          |
| docker.io/mendersoftware/gui:mender-X.Y.Z                    | docker.io/mendersoftware/gui:vX.Y.Z                    |
| docker.io/mendersoftware/inventory:mender-X.Y.Z              | docker.io/mendersoftware/inventory:vX.Y.Z              |
| docker.io/mendersoftware/iot-manager:mender-X.Y.Z            | docker.io/mendersoftware/iot-manager:vX.Y.Z            |
| docker.io/mendersoftware/useradm:mender-X.Y.Z                | docker.io/mendersoftware/useradm:vX.Y.Z                |
| docker.io/mendersoftware/workflows:mender-X.Y.Z              | docker.io/mendersoftware/workflows:vX.Y.Z              |
| docker.io/mendersoftware/workflows-worker:mender-X.Y.Z       | docker.io/mendersoftware/workflows:vX.Y.Z              |
[/ui-tab]
[ui-tab title="Enterprise"]
| From                                                                  | To                                                                        |
| --------------------------------------------------------------------- | ------------------------------------------------------------------------- |
| registry.mender.io/mendersoftware/auditlogs:mender-X.Y.Z              | registry.mender.io/mender-server-enterprise/auditlogs:vX.Y.Z              |
| docker.io/mendersoftware/create-artifact-worker:mender-X.Y.Z          | registry.mender.io/mender-server-enterprise/create-artifact-worker:vX.Y.Z |
| registry.mender.io/mendersoftware/deployments-enterprise:mender-X.Y.Z | registry.mender.io/mender-server-enterprise/deployments:vX.Y.Z            |
| registry.mender.io/mendersoftware/deviceauth-enterprise:mender-X.Y.Z  | registry.mender.io/mender-server-enterprise/deviceauth:vX.Y.Z             |
| docker.io/mendersoftware/deviceconfig:mender-X.Y.Z                    | registry.mender.io/mender-server-enterprise/deviceconfig:vX.Y.Z           |
| docker.io/mendersoftware/deviceconnect:mender-X.Y.Z                   | registry.mender.io/mender-server-enterprise/deviceconnect:vX.Y.Z          |
| registry.mender.io/mendersoftware/devicemonitor:mender-X.Y.Z          | registry.mender.io/mender-server-enterprise/devicemonitor:vX.Y.Z          |
| registry.mender.io/mendersoftware/generate-delta-worker:mender-X.Y.Z  | registry.mender.io/mender-server-enterprise/generate-delta-worker:vX.Y.Z  |
| registry.mender.io/mendersoftware/inventory-enterprise:mender-X.Y.Z   | registry.mender.io/mender-server-enterprise/inventory:vX.Y.Z              |
| docker.io/mendersoftware/iot-manager:mender-X.Y.Z                     | registry.mender.io/mender-server-enterprise/iot-manager:vX.Y.Z            |
| registry.mender.io/mendersoftware/tenantadm:mender-X.Y.Z              | registry.mender.io/mender-server-enterprise/tenantadm:vX.Y.Z              |
| registry.mender.io/mendersoftware/useradm:mender-X.Y.Z                | registry.mender.io/mender-server-enterprise/useradm:vX.Y.Z                |
| registry.mender.io/mendersoftware/workflows:mender-X.Y.Z              | registry.mender.io/mender-server-enterprise/workflows:vX.Y.Z              |
| registry.mender.io/mendersoftware/workflows-worker:mender-X.Y.Z       | registry.mender.io/mender-server-enterprise/workflows:vX.Y.Z              |
[/ui-tab]
[/ui-tabs]

#### Drop the "mender-" prefix in the tags names

> The new image tags drop the `mender-` prefix, so this is no longer valid:

Reviewer: Maybe we should add some context what this change means?
```suggestion
As part of the Mender Server v4 release, we are splitting the Mender Server from the rest of the Mender components (Mender Client, `mender-cli`, `mender-artifact`). Therefore we are dropping the Mender bundle tags `mender-x.y.z`. The new tags are on the form `vX.Y.Z` where X.Y.Z is the *Mender Server* version.
The new image tags drop the `mender-` prefix, so this is no longer valid:
```

```
and reference it in the `mender-values.yaml` file:

```yaml
global:
  # image:
  #   username: "redacted"
  #   password: "redacted"
default:
  imagePullSecrets:
    - name: my-mender-pull-secret
```

#### Rootless gui container
For improved security, the `gui` container is now rootles; this means that
the `gui.httpPort` is switched from `80` to the unprivileged `8090` port.
Make sure you are not overriding the `gui.httpPort` in your
`mender-values.yaml` file.

#### Service Keys automatically generated
You can choose to specify the `device_auth.certs.key`,
`useradm.certs.key`, and `tenantadm.certs.key` keys in
the `mender-values.yaml` file, but it is not mandatory anymore.
If you don't specify them, the Helm Chart will generate them for you.

#### NATS and MongoDB subcharts enabled by default
To facilitate the first Helm Chart installation, the MongoDB and NATS
subcharts are enabled by default:
```yaml
mongodb:
  enabled: true

nats:
  enabled: true
```
If you are using an external MongoDB or NATS instance, please make sure to 
explicitly disable them before the upgrade:
```yaml
mongodb:
  enabled: false

nats:
  enabled: false
```

#### Redis subchart disabled by default

> The Redis subchart is disabled by default, because not used in the Open

Reviewer: ```suggestion
The Redis subchart is disabled by default, because it is not used in the Open
```

Source version. If you want to use it in the Enterprise version, please
make sure to enable it:
```yaml
redis:
  enabled: true
```

#### Enterprise false by default
The Enterprise version is disabled by default. If you are using it, please
make sure to enable it:
```yaml
global:
  enterprise: true
```

#### Storage Proxy enabled by default
By default, the Storage Proxy is enabled. If you don't want to use it, please
restore the previous configuration:
```yaml
api_gateway:
  storage_proxy:
    enabled: false

deployments:
  customEnvs: []
```

But if you want to use it, you have to set the `DEPLOYMENTS_STORAGE_PROXY_URI`
environment variable to the URL exposed, the bucket name in the 
API Gateway configuration, and the `s3.AWS_URI` to the storage endpoint:
```yaml
global:
  s3:
    AWS_URI: "${STORAGE_ENDPOINT}"
    AWS_FORCE_PATH_STYLE: "true"

api_gateway:
  storage_proxy:
    enabled: true
    url: "${STORAGE_ENDPOINT}"

>     customRule: "PathRegexp(\`^/${STORAGE_BUCKET}\`)"

Reviewer: Is this aligned with the helm migration guide?
```suggestion
    customRule: "PathRegexp(`^/${STORAGE_BUCKET}`)"
```

Author: Now it is ;)

    passHostHeader: false

deployments:
  customEnvs:
    - name: DEPLOYMENTS_STORAGE_PROXY_URI
      value: "https://${MENDER_SERVER_DOMAIN}"
```

where:
* `MENDER_SERVER_DOMAIN` is the domain of the Mender Server, e.g:
  `mender.example.com`
* `STORAGE_ENDPOINT` is the endpoint of the storage, e.g: 
  `https://s3.${AWS_REGION}.amazonaws.com`
* `AWS_BUCKET` is the name of the bucket, e.g: `mender-artifacts`

## Maintenance window

> Be prepared for the maintenance window of approximately 15 minutes. This is the maximal

Reviewer: Just double checking, how sure are you about this claim? :mag:

Author: 10%? Let's change it

key to the `default` key. For this reason, the `global.image` key is 
moved to the `default.image` key: please make sure to comment out
your `global.image` key:
<!--AUTOVERSION: "  #   tag: mender-%"/ignore -->
```yaml
global:
  # image:
  #   tag: mender-3.7.7
```

### New repositories location
Following a migration from multiple repositories to a single Monorepo, the
Container repositories has been moved to respect the new structure.
Additionally, the `mender-x.y` tag has been replaced with the `vx.y` tag.
* Enterprise registry:
  1. removing `-enterprise` suffix;
  2. changing `/mendersoftware` to `/mender-server-enterprise`;
  3. moving the remaining components from Docker Hub to `registry.mender.io`.

* Open Source registry:
  1. moving from `docker.io/mendersoftware/deployments:mender-3.7`
  to `docker.io/mendersoftware/deployments:v4.0`.

The new repository structure is already reflected in the default `values.yaml`
file; please make sure to not override it in your `mender-values.yaml` file.

!! If you are mirroring the images, please make sure to update the mirroring
!! configuration to reflect the new repository structure.

> Here's a detailed and comprehensive list of the changes:

Reviewer: ```suggestion
The following table provides a comprehensive list of the changes to the docker image references.
```

[ui-tabs position="top-left" active="0" theme="default" ]
[ui-tab title="Open Source"]

> | From                                                         | To                                                     |

Reviewer: Hmm... Maybe we could improve that header? I'm struggling to come up with something better though
```suggestion
| Former image reference                                               | New image reference                                                     |
```

| docker.io/mendersoftware/deployments:mender-X.Y.Z            | docker.io/mendersoftware/deployments:vX.Y.Z            |
| docker.io/mendersoftware/deviceauth:mender-X.Y.Z             | docker.io/mendersoftware/deviceauth:vX.Y.Z             |
| docker.io/mendersoftware/deviceconfig:mender-X.Y.Z           | docker.io/mendersoftware/deviceconfig:vX.Y.Z           |
| docker.io/mendersoftware/deviceconnect:mender-X.Y.Z          | docker.io/mendersoftware/deviceconnect:vX.Y.Z          |
| docker.io/mendersoftware/gui:mender-X.Y.Z                    | docker.io/mendersoftware/gui:vX.Y.Z                    |
| docker.io/mendersoftware/inventory:mender-X.Y.Z              | docker.io/mendersoftware/inventory:vX.Y.Z              |
| docker.io/mendersoftware/iot-manager:mender-X.Y.Z            | docker.io/mendersoftware/iot-manager:vX.Y.Z            |
| docker.io/mendersoftware/useradm:mender-X.Y.Z                | docker.io/mendersoftware/useradm:vX.Y.Z                |
| docker.io/mendersoftware/workflows:mender-X.Y.Z              | docker.io/mendersoftware/workflows:vX.Y.Z              |

> | docker.io/mendersoftware/workflows-worker:mender-X.Y.Z       | docker.io/mendersoftware/workflows:vX.Y.Z              |

Reviewer: Should we highlight in some way that the `workflows-worker` image is discontinued? Maybe as a footnote?

Same applies to the enterprise table.

Author: I added a warning note: 
![image](https://github.com/user-attachments/assets/fb74055e-e2d8-4f11-a6ff-32fa7c2cca86)

| docker.io/mendersoftware/deviceconnect:mender-X.Y.Z                   | registry.mender.io/mender-server-enterprise/deviceconnect:vX.Y.Z          |
| registry.mender.io/mendersoftware/devicemonitor:mender-X.Y.Z          | registry.mender.io/mender-server-enterprise/devicemonitor:vX.Y.Z          |
| registry.mender.io/mendersoftware/generate-delta-worker:mender-X.Y.Z  | registry.mender.io/mender-server-enterprise/generate-delta-worker:vX.Y.Z  |
| registry.mender.io/mendersoftware/inventory-enterprise:mender-X.Y.Z   | registry.mender.io/mender-server-enterprise/inventory:vX.Y.Z              |
| docker.io/mendersoftware/iot-manager:mender-X.Y.Z                     | registry.mender.io/mender-server-enterprise/iot-manager:vX.Y.Z            |
| registry.mender.io/mendersoftware/tenantadm:mender-X.Y.Z              | registry.mender.io/mender-server-enterprise/tenantadm:vX.Y.Z              |
| registry.mender.io/mendersoftware/useradm:mender-X.Y.Z                | registry.mender.io/mender-server-enterprise/useradm:vX.Y.Z                |
| registry.mender.io/mendersoftware/workflows:mender-X.Y.Z              | registry.mender.io/mender-server-enterprise/workflows:vX.Y.Z              |

> | registry.mender.io/mendersoftware/workflows-worker:mender-X.Y.Z       | registry.mender.io/mender-server-enterprise/workflows:vX.Y.Z              |

Reviewer: Side-note: This table wraps around and becomes quite difficult to read. Is there anything we can do about it?

Author: I don't know, but I don't think so...
